pkg/server/container_create_unix.go

```diff
@@ -148,11 +149,27 @@ func (c *criService) containerSpec(id string, sandboxID string, sandboxPid uint3
 	specOpts = append(specOpts, oci.WithEnv(env))
 
 	securityContext := config.GetLinux().GetSecurityContext()
-	selinuxOpt := securityContext.GetSelinuxOptions()
-	processLabel, mountLabel, err := initSelinuxOpts(selinuxOpt)
+	labelOptions := toLabel(securityContext.GetSelinuxOptions())
+	if len(labelOptions) == 0 {
+		// Use pod level SELinux config
+		if sandbox, err := c.sandboxStore.Get(sandboxID); err == nil {
+			labelOptions, err = label.DupSecOpt(sandbox.ProcessLabel)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
```

Comment:
It's not the current behavior, but something I'd really like is the ability for all containers in a pod to end up with the same MCS pair.

Right now, I can specify user / role / type, and each container will get a different MCS pair. If I specify all the way down to the level, each container will have the same MCS pair.

The consequence of this is that if you have a pod security policy that specifies everything except the level, with the goal of preventing pods from running with a privileged label, you can't also benefit from MCS separation.

This is tantalizingly close to that - we reuse the MCS pair for all containers in a pod - but only if we have no SELinux options at all.

Reply:
Should this inheritance apply to all options user/role/type/level?  So in general if the container did not set the value of user, role, type, or level then copy that value from the pod?  So basically the logic is not actually specific or special for MCS.

Reply:
Potentially yes. It's tempting to apply the merging and inheritance rules here, but I wonder if it would be better handled on the kubelet side, to ensure uniformity.

Today if the pod has a complete set of options and the container has only a partial set, the partial set is sent, and because it's incomplete it gets discarded. kubelet could instead send the most complete set of options available.

The argument for handling all merging and inheritance within the CRI implementation is that some aspects, like the default labels for the system, are only available to it. MCS is also implementation-specific, and cannot be set by the client in a useful or reasonable way, so no matter what we need a way to map the MCS pair onto the specified level (if any). Since we ultimately need to do some amount of inheritance and merging here, it's better to handle all of it, and just specify the expected result for all implementations as part of conformance testing.

I could buy that, but I'm new here and open to being totally wrong.

(My interest is primarily in the mount label side - I need to ensure that _some_ context option is always passed for the rootfs mount - but that needs to match any MCS pair on the process.)

Reply:
Yah I think it should be up to cri to merge the labels at this point.

Altho MCS is implementation specific there is only one implementation used afaik (opencontainers/selinux/go-selinux/label). That implementation depends on there being no other use of MCS on the system except by this library for the auto-creation of categories. We can say that reasonably all MCS implementations should create labels using a set of 2+ categories, since categories are limited to a small range (1024 currently, 256 previously, also see [this article](https://opensource.com/article/18/2/understanding-selinux-labels-container-runtimes)). I have tried to codify what auto-generated labels should look like in the linked cri-tools test PR. That being said I am planning to re-implement the selinux label code for better safety & usability.

Some selinux role/type defaults are defined in `/etc/selinux/targeted/contexts/lxc_contexts`. There are no privileged contexts defined here, instead it is expect to be inherited from the executable's file label or applied through a selinux policy. This is also why I think the container-selinux policy needs to be expanded for containerd.